### ISSUE TITLE
[Source Generator] Source Generator works with dotnet cli but fails with `TypeLoadException` in VS

### ISSUE THREAD

**Issue by LittleLittleCloud** [dotnet/runtime#108164]
### Description

The source generator can successfully generate source code when I compile using dotnet cli. It also successfully generates source code in VS UI when I expand the `analyzer` node. But it raises a `TypeLoadException` when I compile and run my code in VS by hitting `control + F5`.

Here is the link to the source code of source generator: https://github.com/tryAGI/AutoSDK/tree/main/src/libs/AutoSDK.SourceGenerators

Here is the `dotnet --list-sdks`
```bash
8.0.400 [C:\Program Files\dotnet\sdk]
9.0.100-preview.7.24407.12 [C:\Program Files\dotnet\sdk]
```

Here is the error log in VS: 
```
Build started at 9:24 AM...
1>------ Build started: Project: SourceGen, Configuration: Debug Any CPU ------
1>CSC : warning CS8784: Generator 'SdkGenerator' failed to initialize. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'TypeLoadException' with message 'Could not load type 'AutoSDK.Models.ModelData' from assembly 'AutoSDK, Version=0.0.0.0, Culture=neutral, PublicKeyToken=603f13207e65c17b'.'.
1>SourceGen -> C:\Users\xiaoyuz\source\repos\SourceGen\bin\Debug\net8.0\SourceGen.dll
1>Done building project "SourceGen.csproj".
```

Here is the screenshot of which VS I am using:
![image](https://github.com/user-attachments/assets/26216c6a-261e-45f2-9b32-f2d4ba974022)

Finally, the link to a minimal reproducable example for this bug.
- https://github.com/LittleLittleCloud/ItRaiseTypeLoadedExceptionFomSourceGenerator

### Reproduction Steps

See above

### Expected behavior

Source generator successfully generate source in VS

### Actual behavior

It raise a TypeLoadException

**Comment by davidwrighton** [comment]
Yes @elinor-fung appears to have the correct analysis. This is a bug that has been fixed in the runtime, but the fix is far too risky to backport to the .NET Framework.
